### PR TITLE
Achievements recorded on save, not death

### PR DESCRIPTION
**Pull request by tbmo**
#### Summary
Infrastructure "Achievements are now recorded when the game is saved, rather than when your character dies."

#### Purpose of change
I made this change because I like achievements, but I often get bored of a character before they die. So I do not get the achievements recorded, and it doesn't unlock the new scenarios/professions. This would lead to me turning off meta progression. I made this change so that I can get an achievement, save and quit, and immediately go use those professions or scenarios without dying.

#### Describe the solution
Previously, achievements were recorded in player memorials. "past_games_info" reads all of the memorials and parses the events out to figure what achievements you have. However, you only get a memorial when you die. I added a class "past_achievements_info" which has similar functionality but just for achievements. Now, when the game is saved, a list of your achievement_ids is written to save/achievements/{name}-{character id}.json. When the past_achievements_info class is called to check on achievements, if it has not already loaded them, it goes through your save/achievements/ directory to collect all the achievement_ids and uses that to determine eligibility.

Past_achievements_info does duplicate some of the functionality from past_games_info, however, past_games_info has a lot more functionality relating to recording stats, recording which of your characters first completed an achievement, and some other stuff that I didn't touch. I think past_games_info still has a lot of value and I didn't remove anything from it.

#### Describe alternatives you've considered
My first idea was to make the game write a memorial every time the game is saved. I didn't like that because memorials are kind of large and I think they are being used in other places as records of death. I think that might be confusing if some other system is trying to figure out how many times you died and there is like 50 quicksaves of the same character dying.

I also considered making just one global achievements record, a running list of all the achievements that had been attained (so not separate for each character) but I didn't do that because I think a player might want to delete the achievements from some of their characters. That is kind of an edge case but I also didn't like the chance that something could go wrong during the writing process and delete the entire achievement record somehow.

#### Testing
So I would go in with a new character, teleport around to locations that give me achievements, (malls, military bases) and also do debug stuff like go to the bottom of the ocean to get those achievements. Then I would save and quit, make a new character with the professions/scenarios that the previous achievements unlocked, and confirm that I was able to utilize those achievement unlocks.

Something else I did to test specific scenarios was manually edit the achievement entries in save/achievements/ to unlock those achievements that are not as easy to get. Something to note for anyone testing this, if you do this be careful that you also add the prerequisite achievements for some achievements that are hidden or it won't work. In normal gameplay you wouldn't be able to cause this to happen so this would only be a problem for those manually editing their achievement record.

I also tested the arcade mode achievement, which now becomes record on save as it is an achievement id, and confirmed that it's working. Additionally I changed the debug message for unlock all progression to reflect the new way achievements would be handled.

### CODE REVIEW

File: src/scenario.cpp [hunk 1]
```diff
@@ -654,19 +655,15 @@ ret_val<void> scenario::can_afford( const scenario &current_scenario, const int
...
 {
     // if meta progression is disabled then skip this
-    if( get_past_games().achievement( achievement_achievement_arcade_mode ) ||
+    if( get_past_achievements().is_completed( achievement_achievement_arcade_mode ) ||
         !get_option<bool>( "META_PROGRESS" ) ) {
         return ret_val<void>::make_success();
     }
 
     if( _requirement ) {
-        const achievement_completion_info *other_games = get_past_games().achievement(
```
Comment (rtxyd) [inline src/scenario.cpp:663]: I think if delete this part, the achievements in memorial folder can't be record correctly. Maybe we also need to check  the achievements in `const past_games_info &get_past_games()`

Comment (baileytom) [reply src/scenario.cpp:663]: As far as I know this function past_games_info::achievement() just looks for the achievement in the already parsed memorial files and doesn't do any recording. This is just a check to see if its in there

Comment (rtxyd) [reply src/scenario.cpp:663]: yes, the past achievement are all stored in memorial files, when we update game version and move the memorial folder to new game folder, the archievement before this PR will not be checked, so I mean that

Comment (baileytom) [reply src/scenario.cpp:663]: That is true, those memorial achievements would not get migrated over without some specific work

Comment (rtxyd) [reply src/scenario.cpp:663]: That sounds good. In fact, if cdda a commercial game, make changes like clear players' even sunk cost, it 's definitely unacceptable, and players will sue the game designer.

File: src/past_achievements_info.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+#ifndef CATA_SRC_PAST_ACHIEVEMENTS_INFO_H
+#define CATA_SRC_PAST_ACHIEVEMENTS_INFO_H
+
+#include <set>
+
+#include "achievement.h"
+
+struct achievements_completed {
```
Comment (BrettDong) [inline src/past_achievements_info.h:9]: Is this struct used anywhere?

Comment (baileytom) [reply src/past_achievements_info.h:9]: it is not

File: src/scenario.cpp (hunk 1 above)
Comment (baileytom) [reply src/scenario.cpp:663]: Memorial achievements are now being handled

File: src/past_achievements_info.h (hunk 2 above)
Comment (baileytom) [reply src/past_achievements_info.h]: It is removed

**Pull request merged**